pipeline/deploy/cloudformation/codepipeline/cfn_pipeline.yml

```diff
@@ -301,6 +302,236 @@ Resources:
+      Runtime: python3.7
+      Handler: index.handler
+      Timeout: 300
+      Code:
+        ZipFile: |
+          import json
+          import traceback
+          import io
+          from zipfile import ZipFile
+
+          import boto3
+          from botocore.exceptions import ClientError
+
+          code_pipeline = boto3.client("codepipeline")
+          s3 = boto3.client("s3")
+
+          def get_output_prefix(branch_type, stack_name, revision):
+            return f"{branch_type}-{stack_name}-{revision[:12]}"
+
+          def get_source(event):
+            artifacts = event["CodePipeline.job"]["data"]["inputArtifacts"]
+            artifact = [
+              artifact for artifact in artifacts
+              if artifact["name"] == "TemplateSource"
+            ][0]
+            return (
+              artifact["revision"],
+              artifact["location"]["s3Location"]["bucketName"], 
+              artifact["location"]["s3Location"]["objectKey"]
+            )
+
+          def get_user_params(event):
+            raw_params = event["CodePipeline.job"]["data"]["actionConfiguration"]["configuration"]["UserParameters"]
+            params = json.loads(raw_params)
+            return params["BranchType"], params["StackName"], params["TargetBucket"]
+
+          def unpack_source(input_bucket, input_key, target_bucket, target_prefix):
+            response = s3.get_object(Bucket=input_bucket, Key=input_key)
+            archive = ZipFile(io.BytesIO(response["Body"].read()))
+            for name in archive.namelist():
+                target_key = f"{target_prefix}/{name}"
+                print(target_key)
+                s3.put_object(
+                    Body=archive.read(name),
+                    Bucket=target_bucket,
+                    Key=target_key
+                )
+
+          def handler(event, context):
+            try:
+              job_id = event["CodePipeline.job"]["id"]
+              print(event)
+              
+              revision, bucket, key = get_source(event)
+              branch_type, stack_name, target_bucket = get_user_params(event)
+
+              prefix = get_output_prefix(branch_type, stack_name, revision)
+              unpack_source(bucket, key, target_bucket, prefix)
+
+              code_pipeline.put_job_success_result(jobId=job_id)
+
+            except Exception as err:
+              print(err)
+              traceback.print_exc()
+              code_pipeline.put_job_failure_result(
+                jobId=job_id, 
+                failureDetails={'message': str(err), 'type': 'JobFailed'}
+              )
+              raise
+
+  # Setup variables for the pipeline build
+  WriteDynamicParameters:
+    Type: AWS::Lambda::Function
+    Properties:
+      Role: !GetAtt PrepareStackForDeletionLambdaRole.Arn
+      Runtime: python3.7
+      Handler: index.handler
+      Timeout: 300
+      Code:
+        ZipFile: |
+          import json
```

Comment:
I think the code statement can only refer to [a deployment package in s3](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-code.html) or be written inline. The former is definitely better, but I don't know how to apply it here without complicating our deployment process. 

(This is not the case for lambdas used by the morphology_pipeline itself, since they can refer to the deployment package built by Codepipeline.)